lib/install/deps.js

```diff
@@ -643,8 +643,21 @@ var earliestInstallable = exports.earliestInstallable = function (requiredBy, tr
   // if this tree location requested the same module then we KNOW it
   // isn't compatible because if it were findRequirement would have
   // found that version.
-  var deps = tree.package.dependencies || {}
-  if (!tree.removed && requiredBy !== tree && deps[pkg.name]) {
+  var deps = Object.keys(tree.package.dependencies || {})
+
+  // if we're at the root of the tree, and are planning to install
+  // devDependencies in a future pass, don't put any conflicting
+  // packages in the root during the first pass.
+  if (!tree.parent && (npm.config.get('dev') || !npm.config.get('production'))) {
```

Comment:
We also have `--also` and `--only`. This should probably be:

```
var alwaysIncludeDev = npm.config.get('dev') || /^dev(elopment)$/.test(npm.config.get('also'))
var onlyIncludeProduction = npm.config.get('production') || /^prod(uction)$/.test(npm.config.get('only'))
if (!tree.parent && (alwaysIncludeDev || !onlyIncludeProduction)) {
```

But to be honest, I'm not sure we should even gate this? Why not do this kind of hoisting even if you ask for `--only=prod`? It'd ensure that you got the same tree with `npm install --only=prod && npm install --only=dev` as you would with `npm install`